File: cloud/src/core/stats.ts
/**
 * Phase 3.3 §2.2-§2.4 / §2.6: 每台 worker 跑每種工作有多快。
 * Parity source: `server/comfyfed_server/stats.py` -- 上半是純函數（逐行對
 * 照），下半是薄薄的 D1 adapter。統計壞掉絕不能影響 job_done 主流程與收據
 * （spec §5），所以每個 adapter 都自己吞例外。
 */

import * as queries from "../db/queries";
import { toSqliteTimestamp } from "../db/queries";
import { extract, signature as computeSignature } from "./assess";
import { speedPrior } from "./scheduler";

export const EWMA_ALPHA = 0.3;
export const SPEED_MIN = 0.1;
export const SPEED_MAX = 10.0;
export const DEFAULT_PREDICTED_SECONDS = 60.0;

export const BACKFILL_LIMIT = 500;
export const BACKFILL_SETTING_KEY = "stats_backfilled";

export type Basis = "signature" | "speed_index" | "fleet_default" | "none";

export interface StatRow {
  workerId: string;
  signature: string;
  ewmaSeconds: number;
  samples: number;
}

export function isValidExecSeconds(execSeconds: unknown): execSeconds is number {
  return typeof execSeconds === "number" && Number.isFinite(execSeconds) && execSeconds >= 0;
}

// --- 純函數層 --------------------------------------------------------------

/** 偶數個取中間兩個的平均，空的回 null -- ports `stats.median`. */
export function median(values: number[]): number | null {
  if (values.length === 0) return null;
  const ordered = [...values].sort((a, b) => a - b);
  const mid = Math.floor(ordered.length / 2);
  if (ordered.length % 2 === 1) return ordered[mid]!;
  return (ordered[mid - 1]! + ordered[mid]!) / 2;
}

/** §2.3 的 R(sig) -- ports `stats.fleet_reference`. */
export function fleetReference(
  rows: StatRow[],
  speedIndex: Map<string, number>,
  signature: string,
  excludeWorkerId: string | null
): number | null {
  const values = rows
    .filter((r) => r.signature === signature && r.workerId !== excludeWorkerId)
    .map((r) => r.ewmaSeconds * (speedIndex.get(r.workerId) ?? 1.0));
  return median(values);
}

/** Ports `stats.next_ewma`. */
export function nextEwma(previous: number | null, execSeconds: number): number {
  if (previous === null) return execSeconds;
  return EWMA_ALPHA * execSeconds + (1 - EWMA_ALPHA) * previous;
}

/** Ports `stats.next_speed_index` -- 沒有參考值或 exec<=0 一律維持原值。 */
export function nextSpeedIndex(current: number, reference: number | null, execSeconds: number): number {
  if (reference === null || execSeconds <= 0) return current;
  const ratio = reference / execSeconds;
  const blended = EWMA_ALPHA * ratio + (1 - EWMA_ALPHA) * current;
  return Math.max(SPEED_MIN, Math.min(SPEED_MAX, blended));
}

/** First completed job of a worker: `R(sig)/exec` clamped, or the fallback
 * (its prior) when there is no reference to compare against -- ports
 * `stats.first_speed_index`. */
export function firstSpeedIndex(reference: number | null, execSeconds: number, fallback: number): number {
  if (reference === null || execSeconds <= 0) return fallback;
  return Math.max(SPEED_MIN, Math.min(SPEED_MAX, reference / execSeconds));
}

/** 2026-09-23 §2 -- ports `stats.effective_speed_index`: a worker's
 * `speed_index` for prediction purposes. The stored value is learned from
 * completed jobs; a worker with no `worker_job_stats` row at all has learned
 * nothing yet, and its stored 1.0 default would make a Mac look like an
 * NVIDIA card. Such a worker gets `scheduler.speedPrior(backend)` instead. */
export function effectiveSpeedIndex(
  rows: StatRow[],
  workerId: string,
  backend: string,
  stored: number | null | undefined
): number {
  if (rows.some((row) => row.workerId === workerId)) {
    if (typeof stored === "number" && stored > 0) return stored;
    return 1.0;
  }
  return speedPrior(backend);
}

/** §2.4 的四階梯 -- ports `stats.predict`. */
export function predict(
  rows: StatRow[],
  speedIndex: Map<string, number>,
  signature: string | null,
  workerId: string
): { seconds: number; basis: Basis } {
  let ownSpeed = speedIndex.get(workerId) ?? 1.0;
  if (!(ownSpeed > 0)) ownSpeed = 1.0;

  if (signature) {
    for (const row of rows) {
      if (row.workerId === workerId && row.signature === signature) {
        return { seconds: row.ewmaSeconds, basis: "signature" };
      }
    }
    const reference = fleetReference(rows, speedIndex, signature, workerId);
    if (reference !== null) return { seconds: reference / ownSpeed, basis: "speed_index" };
  }

  const fleetMedian = median(rows.map((r) => r.ewmaSeconds));
  if (fleetMedian !== null) return { seconds: fleetMedian / ownSpeed, basis: "fleet_default" };

  return { seconds: DEFAULT_PREDICTED_SECONDS, basis: "none" };
}

// --- D1 adapter 層 ---------------------------------------------------------

export async function loadRows(db: D1Database): Promise<StatRow[]> {
  try {
    const { results } = await db
      .prepare("SELECT worker_id, signature, ewma_seconds, samples FROM worker_job_stats")
      .all<{ worker_id: string; signature: string; ewma_seconds: number; samples: number }>();
    return results.map((r) => ({
      workerId: r.worker_id,
      signature: r.signature,
      ewmaSeconds: r.ewma_seconds,
      samples: r.samples,
    }));
  } catch (err) {
    console.error("stats: loadRows failed", err);
    return [];
  }
}

export async function loadSpeedIndex(db: D1Database): Promise<Map<string, number>> {
  try {
    const { results } = await db
      .prepare("SELECT id, speed_index FROM workers")
      .all<{ id: string; speed_index: number }>();
    return new Map(results.map((r) => [r.id, typeof r.speed_index === "number" ? r.speed_index : 1.0]));
  } catch (err) {
    console.error("stats: loadSpeedIndex failed", err);
    return new Map();
  }
}

/** 2026-09-23 §2: the `speed_index` map `record_completion` / the backfill
 * build in Python -- `effective_speed_index(rows, w.id, w.backend, w.speed_index)`
 * over every worker row. Same swallow-and-empty contract as `loadSpeedIndex`. */
export async function loadEffectiveSpeedIndex(db: D1Database, rows: StatRow[]): Promise<Map<string, number>> {
  try {
    const { results } = await db
      .prepare("SELECT id, backend, speed_index FROM workers")
      .all<{ id: string; backend: string | null; speed_index: number | null }>();
    return new Map(results.map((r) => [r.id, effectiveSpeedIndex(rows, r.id, r.backend || "", r.speed_index)]));
  } catch (err) {
    console.error("stats: loadEffectiveSpeedIndex failed", err);
    return new Map();
  }
}

/** §2.3 -- ports `stats.record_completion`. */
export async function recordCompletion(
  db: D1Database,
  workerId: string,
  signature: string | null,
  execSeconds: number | null,
  now: Date
): Promise<void> {
  if (!signature || !isValidExecSeconds(execSeconds)) return;

  try {
    const rows = await loadRows(db);
    const speedIndex = await loadEffectiveSpeedIndex(db, rows);
    const firstSample = !rows.some((r) => r.workerId === workerId);

    const existing = rows.find((r) => r.workerId === workerId && r.signature === signature) ?? null;
    const ewma = nextEwma(existing ? existing.ewmaSeconds : null, execSeconds);
    const samples = (existing?.samples ?? 0) + 1;

    await db
      .prepare(
        `INSERT INTO worker_job_stats (worker_id, signature, ewma_seconds, samples, updated_at)
         VALUES (?, ?, ?, ?, ?)
         ON CONFLICT (worker_id, signature) DO UPDATE SET
           ewma_seconds = excluded.ewma_seconds,
           samples = excluded.samples,
           updated_at = excluded.updated_at`
      )
      .bind(workerId, signature, ewma, samples, toSqliteTimestamp(now))
      .run();

    // 參考值算在 EWMA 更新「之前」的快照（`rows`）上，且排除自己 -- 否則這
    // 一筆會同時當觀測值和參考值，自我校正成 1.0。
    const reference = fleetReference(rows, speedIndex, signature, workerId);
    const current = speedIndex.get(workerId) ?? 1.0;
    // 2026-09-23 §2: the very first sample REPLACES the prior instead of
    // blending 30 % into it -- same rule as the EWMA's first sample. Blending
    // would leave a Mac at 0.7 after one job and take ten more to converge.
    // Python writes `worker.speed_index` unconditionally (when the row
    // exists); here the UPDATE is a no-op on an unknown worker, so it is
    // issued whenever the map knew the worker (its prior may differ from the
    // stored 1.0 even when nothing was learned).
    const updated = firstSample
      ? firstSpeedIndex(reference, execSeconds, current)
      : nextSpeedIndex(current, reference, execSeconds);
    if (speedIndex.has(workerId)) {
      await db.prepare("UPDATE workers SET speed_index = ? WHERE id = ?").bind(updated, workerId).run();
    }
  } catch (err) {
    console.error(`stats: recordCompletion failed for worker ${workerId} signature ${signature}`, err);
  }
}

/** §2.6 -- ports `stats.backfill_if_needed`；在 Hub DO 第一次 tick 時呼叫。
 *
 * Final-review I3：舊版本在迴圈裡逐筆呼叫 `recordCompletion`，每一筆都重讀
 * 一整張 `worker_job_stats` 加一整張 `workers`，再寫回去 -- 500 筆就是 500
 * 次全表讀 + 1000 次寫，全在 DO alarm 裡。現在改成：讀一次、在記憶體裡用
 * 同一組純函數（`nextEwma` / `nextSpeedIndex` / `fleetReference`）依序重放，
 * 最後一次 `db.batch` 把結果寫出去。重放的順序、快照時點（參考值算在
 * EWMA 更新「之前」的快照上、且排除自己）跟 `recordCompletion` 逐行一致，
 * 所以 N 筆回填的結果等於 N 次順序 `recordCompletion`。
 *
 * 旗標只在寫入成功之後才設；這裡不再吞例外，讓它丟回 `hub.ts`，
 * 那邊會把 per-instance 的 `statsBackfillDone` 留在 false，下一個 tick 再試。 */
export async function backfillIfNeeded(db: D1Database): Promise<boolean> {
  if ((await queries.getSetting(db, BACKFILL_SETTING_KEY)) === "1") return false;

  const any = await db.prepare("SELECT 1 AS present FROM worker_job_stats LIMIT 1").first<{ present: number }>();
  if (any) {
    await queries.setSetting(db, BACKFILL_SETTING_KEY, "1");
    return false;
  }

  const { results } = await db
    .prepare(
      `SELECT r.worker_id AS worker_id, r.gpu_seconds AS gpu_seconds, r.created_at AS created_at,
              j.id AS job_id, j.signature AS signature, j.workflow_json AS workflow_json
       FROM receipts r JOIN jobs j ON j.id = r.job_id
       WHERE r.kind = 'completed' AND r.billable = 1
       ORDER BY r.created_at DESC
       LIMIT ?`
    )
    .bind(BACKFILL_LIMIT)
    .all<{
      worker_id: string;
      gpu_seconds: number;
      created_at: string;
      job_id: string;
      signature: string | null;
      workflow_json: string;
    }>();

  const replay = [...results].reverse(); // 依 created_at 由舊到新

  // 讀一次，全程在記憶體裡演進。`worker_job_stats` 這時一定是空的
  // （上面的 `any` 檢查已經提前返回），還是照讀一次保持與
  // `recordCompletion` 同形。
  const rowsMap = new Map<string, StatRow & { updatedAt: Date }>();
  // 2026-09-23 §2：和 `recordCompletion` 一樣，沒有任何一列的 worker 用後端
  // 先驗；有列的用存的值。重放途中第一筆樣本出現時改用 `firstSpeedIndex`
  // （取代先驗而不是混合）。
  const speedIndex = await loadEffectiveSpeedIndex(db, [...rowsMap.values()]);
  // `loadEffectiveSpeedIndex` 只列真實存在的 worker。收據裡的 worker 已被刪除
  // 時，`recordCompletion` 算出來的 UPDATE 是空轉（沒有那一列），而且下一筆
  // 重讀時也拿不到它 -- 所以這裡也只替已知 worker 算新值（同 Python 的
  // `known_workers` 閘門）。
  const knownWorkers = new Set(speedIndex.keys());
  const signatureWrites = new Map<string, string>();
  const speedWrites = new Map<string, number>();

  for (const row of replay) {
    let signature = row.signature;
    if (!signature) {
      let workflow: unknown;
      try {
        workflow = JSON.parse(row.workflow_json || "{}");
      } catch {
        continue;
      }
      if (typeof workflow !== "object" || workflow === null) continue;
      const wf = workflow as Record<string, unknown>;
      signature = await computeSignature(wf, extract(wf));
      signatureWrites.set(row.job_id, signature);
    }
    // `gpu_seconds` 是回填唯一能拿到的執行秒數代理值（收據沒存原始
    // exec_seconds；它本身就是 min(exec_seconds, wall)）。
    const execSeconds = row.gpu_seconds;
    // `recordCompletion` 的兩道前置閘門，一字不漏地搬過來。
    if (!signature || !isValidExecSeconds(execSeconds)) continue;

    const snapshot = [...rowsMap.values()];
    const key = `${row.worker_id} ${signature}`;
    const existing = rowsMap.get(key) ?? null;
    rowsMap.set(key, {
      workerId: row.worker_id,
      signature,
      ewmaSeconds: nextEwma(existing ? existing.ewmaSeconds : null, execSeconds),
      samples: (existing?.samples ?? 0) + 1,
      updatedAt: new Date(`${row.created_at}Z`),
    });

    // 參考值算在 EWMA 更新「之前」的快照上，且排除自己。
    if (knownWorkers.has(row.worker_id)) {
      const reference = fleetReference(snapshot, speedIndex, signature, row.worker_id);
      const current = speedIndex.get(row.worker_id)!;
      const firstSample = !snapshot.some((r) => r.workerId === row.worker_id);
      const updated = firstSample
        ? firstSpeedIndex(reference, execSeconds, current)
        : nextSpeedIndex(current, reference, execSeconds);
      speedIndex.set(row.worker_id, updated);
      speedWrites.set(row.worker_id, updated);
    }
  }

  const statements: D1PreparedStatement[] = [];
  for (const [jobId, sig] of signatureWrites) {
    statements.push(db.prepare("UPDATE jobs SET signature = ? WHERE id = ?").bind(sig, jobId));
  }
  for (const row of rowsMap.values()) {
    statements.push(
      db
        .prepare(
          `INSERT INTO worker_job_stats (worker_id, signature, ewma_seconds, samples, updated_at)
           VALUES (?, ?, ?, ?, ?)
           ON CONFLICT (worker_id, signature) DO UPDATE SET
             ewma_seconds = excluded.ewma_seconds,
             samples = excluded.samples,
             updated_at = excluded.updated_at`
        )
        .bind(row.workerId, row.signature, row.ewmaSeconds, row.samples, toSqliteTimestamp(row.updatedAt))
    );
  }
  for (const [workerId, value] of speedWrites) {
    statements.push(db.prepare("UPDATE workers SET speed_index = ? WHERE id = ?").bind(value, workerId));
  }
  // 上限：最多 BACKFILL_LIMIT(500) 筆簽章補寫 + 500 列統計 + 機隊台數的
  // speed_index，實務上是幾百條序列。
  if (statements.length > 0) await db.batch(statements);

  // 旗標只在寫入成功之後才設。
  await queries.setSetting(db, BACKFILL_SETTING_KEY, "1");
  return true;
}

File: server/comfyfed_server/stats.py
"""Phase 3.3 §2.2-§2.4 / §2.6: 每台 worker 跑每種工作有多快。

模組分兩層：上半是完全不碰 DB 的純函數（EWMA、隊伍參考值、speed_index
更新、predict 的 basis ladder），兩棧逐行對照 `cloud/src/core/stats.ts`；
下半是薄薄的 DB adapter。所有 adapter 都吞掉例外只記 log -- 統計壞掉絕不能
影響 job_done 主流程與收據（spec §5）。
"""

from __future__ import annotations

import json
import logging
import math
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional

from . import assess, db

logger = logging.getLogger(__name__)

EWMA_ALPHA = 0.3
SPEED_MIN = 0.1
SPEED_MAX = 10.0
DEFAULT_PREDICTED_SECONDS = 60.0

# §2.6: 回填只重放最近這麼多筆收據，且只做一次。
BACKFILL_LIMIT = 500
BACKFILL_SETTING_KEY = "stats_backfilled"


@dataclass(frozen=True)
class StatRow:
    worker_id: str
    signature: str
    ewma_seconds: float
    samples: int


def _utcnow() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def is_valid_exec_seconds(exec_seconds) -> bool:
    """有效樣本：真實、有限、非負的數字。和 `agentws._is_valid_exec_seconds`
    同一組條件，複寫在這裡而不是 import，是為了讓本模組維持可獨立測試的純
    函數層（agentws 會拉進整個 WebSocket 世界）。"""
    return (
        exec_seconds is not None
        and isinstance(exec_seconds, (int, float))
        and not isinstance(exec_seconds, bool)
        and math.isfinite(exec_seconds)
        and exec_seconds >= 0
    )


# --- 純函數層 --------------------------------------------------------------


def median(values: list[float]) -> Optional[float]:
    """偶數個取中間兩個的平均，空的回 None。自己寫而不是用 statistics.median
    是為了和 TS 端逐行對照（TS 沒有標準庫中位數）。"""
    if not values:
        return None
    ordered = sorted(values)
    mid = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return ordered[mid]
    return (ordered[mid - 1] + ordered[mid]) / 2.0


def fleet_reference(
    rows: list[StatRow],
    speed_index: dict[str, float],
    signature: str,
    exclude_worker_id: Optional[str],
) -> Optional[float]:
    """§2.3 的 R(sig)：其他 worker 該簽章 `ewma × speed_index` 的中位數。

    乘上 speed_index 是把每台機的觀測值折算回「平均機」的尺度，這樣不同速度
    的機器才能放進同一個中位數比較。沒有其他 worker 的資料時回 None，呼叫端
    據此決定「不更新」或「往下一階 fallback」。
    """
    values = [
        row.ewma_seconds * speed_index.get(row.worker_id, 1.0)
        for row in rows
        if row.signature == signature and row.worker_id != exclude_worker_id
    ]
    return median(values)


def next_ewma(previous: Optional[float], exec_seconds: float) -> float:
    """第一筆樣本直接就是 EWMA 本身；之後 `0.3*exec + 0.7*previous`。"""
    if previous is None:
        return float(exec_seconds)
    return EWMA_ALPHA * float(exec_seconds) + (1.0 - EWMA_ALPHA) * float(previous)


def next_speed_index(current: float, reference: Optional[float], exec_seconds: float) -> float:
    """§2.3 第 2 條：`ratio = R(sig)/exec`，再和現值做同一個 α 的混合後夾住。

    `reference is None`（沒有別人的資料）或 `exec_seconds <= 0`（除不下去）
    一律維持原值 -- 寧可不更新，也不要寫進一個 inf。
    """
    if reference is None or exec_seconds <= 0:
        return current
    ratio = reference / exec_seconds
    blended = EWMA_ALPHA * ratio + (1.0 - EWMA_ALPHA) * current
    return max(SPEED_MIN, min(SPEED_MAX, blended))


def predict(
    rows: list[StatRow],
    speed_index: dict[str, float],
    signature: Optional[str],
    worker_id: str,
) -> tuple[float, str]:
    """§2.4 的四階梯：signature -> speed_index -> fleet_default -> none。

    回傳 `(seconds, basis)`；`basis` 原樣寫進 `jobs.dispatch_info`，console
    就能說明「這個預估是怎麼來的」。
    """
    own_speed = speed_index.get(worker_id, 1.0)
    if not isinstance(own_speed, (int, float)) or own_speed <= 0:
        own_speed = 1.0

    if signature:
        for row in rows:
            if row.worker_id == worker_id and row.signature == signature:
                return row.ewma_seconds, "signature"

        reference = fleet_reference(rows, speed_index, signature, exclude_worker_id=worker_id)
        if reference is not None:
            return reference / own_speed, "speed_index"

    fleet_median = median([row.ewma_seconds for row in rows])
    if fleet_median is not None:
        return fleet_median / own_speed, "fleet_default"

    return DEFAULT_PREDICTED_SECONDS, "none"


# --- DB adapter 層 ---------------------------------------------------------


def load_rows() -> list[StatRow]:
    """整張 `worker_job_stats`。故意一次全讀：一個家用聯邦的 (worker,
    signature) 組合是幾十到幾百列，每個 tick 一次全表讀遠比每對候選跑一次
    查詢便宜，而且讓 `predict` 維持純函數。"""
    try:
        with db.get_session() as session:
            return [
                StatRow(
                    worker_id=row.worker_id,
                    signature=row.signature,
                    ewma_seconds=row.ewma_seconds,
                    samples=row.samples,
                )
                for row in session.query(db.WorkerJobStats).all()
            ]
    except Exception:
        logger.exception("stats: load_rows failed")
        return []


def load_speed_index() -> dict[str, float]:
    try:
        with db.get_session() as session:
            return {
                w.id: (w.speed_index if isinstance(w.speed_index, (int, float)) else 1.0)
                for w in session.query(db.Worker).all()
            }
    except Exception:
        logger.exception("stats: load_speed_index failed")
        return {}


def record_completion(
    worker_id: str, signature: Optional[str], exec_seconds: Optional[float]
) -> None:
    """§2.3：只在 job_done 且 `exec_seconds` 有效時更新。

    失敗只記 log -- 呼叫端（`agentws._handle_job_done`）絕不能因此少發一張
    收據。
    """
    if not signature or not is_valid_exec_seconds(exec_seconds):
        return
    exec_value = float(exec_seconds)

    try:
        with db.get_session() as session:
            rows = [
                StatRow(
                    worker_id=r.worker_id,
                    signature=r.signature,
                    ewma_seconds=r.ewma_seconds,
                    samples=r.samples,
                )
                for r in session.query(db.WorkerJobStats).all()
            ]
            speed_index = {
                w.id: (w.speed_index if isinstance(w.speed_index, (int, float)) else 1.0)
                for w in session.query(db.Worker).all()
            }

            # 1) EWMA。
            existing = session.get(db.WorkerJobStats, (worker_id, signature))
            if existing is None:
                session.add(
                    db.WorkerJobStats(
                        worker_id=worker_id,
                        signature=signature,
                        ewma_seconds=next_ewma(None, exec_value),
                        samples=1,
                        updated_at=_utcnow(),
                    )
                )
            else:
                existing.ewma_seconds = next_ewma(existing.ewma_seconds, exec_value)
                existing.samples = existing.samples + 1
                existing.updated_at = _utcnow()

            # 2) speed_index。參考值算在 EWMA 更新「之前」的快照上，且排除
            #    自己 -- 否則這一筆會同時當觀測值和參考值，自我校正成 1.0。
            worker = session.get(db.Worker, worker_id)
            if worker is not None:
                reference = fleet_reference(
                    rows, speed_index, signature, exclude_worker_id=worker_id
                )
                current = worker.speed_index if isinstance(worker.speed_index, (int, float)) else 1.0
                worker.speed_index = next_speed_index(current, reference, exec_value)

            session.commit()
    except Exception:
        logger.exception(
            "stats: record_completion failed for worker %s signature %s", worker_id, signature
        )


def _apply_backfill_writes(
    session,
    stats_map: dict[tuple[str, str], tuple[float, int, datetime]],
    speed_updates: dict[str, float],
) -> None:
    """回填的寫入階段（Final-review I3）。

    單獨拆出來一個函數，是為了讓「寫入失敗 -> 旗標不設」路徑測得到，
    也讓重放（純計算）跟寫入在閱讀上分開。不自己 commit -- 由呼叫端連同
    旗標一起 commit，這樣任何一步丟例外都不會留下「旗標設了但數據沒寫」
    的狀態。
    """
    for (worker_id, signature), (ewma, samples, updated_at) in stats_map.items():
        existing = session.get(db.WorkerJobStats, (worker_id, signature))
        if existing is None:
            session.add(
                db.WorkerJobStats(
                    worker_id=worker_id,
                    signature=signature,
                    ewma_seconds=ewma,
                    samples=samples,
                    updated_at=updated_at,
                )
            )
        else:
            existing.ewma_seconds = ewma
            existing.samples = samples
            existing.updated_at = updated_at

    for worker_id, value in speed_updates.items():
        worker = session.get(db.Worker, worker_id)
        if worker is not None:
            worker.speed_index = value


def backfill_if_needed() -> bool:
    """§2.6：第一次啟動時用最近 500 筆完成收據重放一次 `record_completion`。

    回傳「這次有沒有真的跑回填」（旗標已設 -> False）。簽章缺的 job 先由
    `workflow_json` 補算並寫回 `jobs.signature`，所以回填同時也是舊資料的
    簽章補齊通道。

    Final-review I3：舊版本在迴圈裡逐筆呼叫 `record_completion`，每一筆都開一
    個 session、重讀一整張 `worker_job_stats` 加一整張 `workers`，再 commit 一次
    -- 500 筆就是 500 輪。現在改成：讀一次、在記憶體裡用同一組純函數
    （`next_ewma` / `next_speed_index` / `fleet_reference`）依序重放，最後在同一個
    session 裡寫出去。重放的順序、快照時點（參考值算在 EWMA 更新「之前」
    的快照上、且排除自己）跟 `record_completion` 逐行一致，所以 N 筆回填的
    結果等於 N 次順序 `record_completion`。

    `stats_backfilled` 旗標跟資料在同一次 commit 裡寫——寫入丟例外就什麼都
    沒進去，旗標也不會被設，下一次啟動會完整重試。
    """
    try:
        with db.get_session() as session:
            flag = session.get(db.Setting, BACKFILL_SETTING_KEY)
            if flag is not None and flag.value == "1":
                return False
            if session.query(db.WorkerJobStats).first() is not None:
                session.add(db.Setting(key=BACKFILL_SETTING_KEY, value="1"))
                session.commit()
                return False

            receipts = (
                session.query(db.Receipt)
                .filter(db.Receipt.kind == "completed", db.Receipt.billable == True)  # noqa: E712
                .order_by(db.Receipt.created_at.desc())
                .limit(BACKFILL_LIMIT)
                .all()
            )
            replay = list(reversed(receipts))  # 依 created_at 由舊到新

            # 讀一次，全程在記憶體裡演進。`worker_job_stats` 這時一定是空
            # 的（上面的檢查已提前返回），還是照讀一次保持與
            # `record_completion` 同形。
            stats_map: dict[tuple[str, str], tuple[float, int, datetime]] = {
                (r.worker_id, r.signature): (r.ewma_seconds, r.samples, r.updated_at)
                for r in session.query(db.WorkerJobStats).all()
            }
            speed_index = {
                w.id: (w.speed_index if isinstance(w.speed_index, (int, float)) else 1.0)
                for w in session.query(db.Worker).all()
            }
            # `record_completion` 只在 worker 列還在時寫 speed_index（且下一筆
            # 重讀時也拿不到已刪除的 worker），這裡同步這個行為。
            known_workers = set(speed_index)
            speed_updates: dict[str, float] = {}

            for receipt in replay:
                if receipt.job_id is None:
                    continue
                job = session.get(db.Job, receipt.job_id)
                if job is None:
                    continue
                signature = job.signature
                if not signature:
                    try:
                        workflow = json.loads(job.workflow_json or "{}")
                    except (TypeError, ValueError):
                        continue
                    if not isinstance(workflow, dict):
                        continue
                    signature = assess.signature(workflow, assess.needs_from_job(job))
                    job.signature = signature
                # `gpu_seconds` 是回填唯一能拿到的執行秒數代理值（收據沒
                # 存原始 exec_seconds；它本身就是 min(exec_seconds, wall)）。
                exec_seconds = receipt.gpu_seconds
                # `record_completion` 的兩道前置閘門，一字不漏地搬過來。
                if not signature or not is_valid_exec_seconds(exec_seconds):
                    continue
                exec_value = float(exec_seconds)

                snapshot = [
                    StatRow(
                        worker_id=key[0],
                        signature=key[1],
                        ewma_seconds=value[0],
                        samples=value[1],
                    )
                    for key, value in stats_map.items()
                ]
                key = (receipt.worker_id, signature)
                previous = stats_map.get(key)
                stats_map[key] = (
                    next_ewma(previous[0] if previous is not None else None, exec_value),
                    (previous[1] if previous is not None else 0) + 1,
                    receipt.created_at or _utcnow(),
                )

                # 參考值算在 EWMA 更新「之前」的快照上，且排除自己。
                if receipt.worker_id in known_workers:
                    reference = fleet_reference(
                        snapshot, speed_index, signature, exclude_worker_id=receipt.worker_id
                    )
                    current = speed_index[receipt.worker_id]
                    updated = next_speed_index(current, reference, exec_value)
                    speed_index[receipt.worker_id] = updated
                    speed_updates[receipt.worker_id] = updated

            _apply_backfill_writes(session, stats_map, speed_updates)

            # 旗標跟資料同一次 commit。
            if flag is None:
                session.add(db.Setting(key=BACKFILL_SETTING_KEY, value="1"))
            else:
                flag.value = "1"
            session.commit()
    except Exception:
        logger.exception("stats: backfill failed")
        return False

    return True
